src/secret.rs

```diff
@@ -0,0 +1,78 @@
+use std::{fmt, str::FromStr};
+use url::Url;
+
+#[derive(Clone, Eq, PartialEq)]
+pub struct SecretUrl(Url);
+
+impl SecretUrl {
+    pub fn new(url: Url) -> Self {
+        Self(url)
+    }
+
+    pub fn expose(&self) -> &str {
+        self.as_ref()
+    }
+
+    fn format(&self) -> String {
+        if self.0.has_authority() {
+            let mut url = self.0.clone();
+            if url.password().is_some() {
+                url.set_password(None).unwrap();
+            }
+            url.set_username("**********").unwrap();
+            return url.to_string();
+        }
+        self.0.to_string()
+    }
+}
+
+impl AsRef<str> for SecretUrl {
+    fn as_ref(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+impl FromStr for SecretUrl {
+    type Err = <Url as FromStr>::Err;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Url::from_str(s).map(SecretUrl::new)
+    }
+}
+
+impl fmt::Display for SecretUrl {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(&self.format())
+    }
+}
+
+impl fmt::Debug for SecretUrl {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(&self.format())
+    }
+}
```

Comment:
Not that I'm a big fan of the debug format of the `Url` type but I feel like it'd be best if we could preserve the different behaviours of debug and display for SecretUrl

Could we make it so that `.format()` returns `Url` with the username and password redacted but still use the `Url` impls of `Debug` and `Display`?